### PR TITLE
feat: add annotation to ingress cr for the clusters above 4.10

### PR DESCRIPTION
**Pull request by boranx**
Fix: https://issues.redhat.com/browse/OSD-10213

_Testing_:

**4.10.3**

```
{"level":"info","ts":1647519636.549309,"logger":"controller_publishingstrategy","msg":"IngressController's CR of apps2 is being patched for the missing annotation: ingress.operator.openshift.io/auto-delete-load-balancer","Request.Namespace":"openshift-cloud-ingress-operator","Request.Name":"publishingstrategy"}
{"level":"info","ts":1647519636.663991,"logger":"controller_publishingstrategy","msg":"IngressController's CR of default is being patched for the missing annotation: ingress.operator.openshift.io/auto-delete-load-balancer","Request.Namespace":"openshift-cloud-ingress-operator","Request.Name":"publishingstrategy"}
```
```shell
oc get ingresscontroller -n openshift-ingress-operator -o yaml | grep "kind: IngressController" -A5
  kind: IngressController
  metadata:
    annotations:
      Owner: cloud-ingress-operator
      ingress.operator.openshift.io/auto-delete-load-balancer: ""
    creationTimestamp: "2022-03-17T12:20:36Z"
--
  kind: IngressController
  metadata:
    annotations:
      Owner: cloud-ingress-operator
      ingress.operator.openshift.io/auto-delete-load-balancer: ""
    creationTimestamp: "2022-03-17T12:18:34Z"
```

**4.9.15**

```shell
oc get ingresscontroller -n openshift-ingress-operator -o yaml | grep "annotations"
# none
```
```shell
oc get ingresscontroller -n openshift-ingress-operator -o yaml | grep "kind: IngressController" -A10
  kind: IngressController
  metadata:
    creationTimestamp: "2022-03-15T14:40:07Z"
    finalizers:
    - ingresscontroller.operator.openshift.io/finalizer-ingresscontroller
    generation: 4
    name: default
    namespace: openshift-ingress-operator
    resourceVersion: "63704"
    uid: 227ec203-d8c6-467b-b03d-c7c3bfb23c56
  spec:
```

### CODE REVIEW

File: pkg/controller/publishingstrategy/publishingstrategy_controller.go [hunk 1]
```diff
@@ -44,6 +44,7 @@ type patchField string
 var IngressControllerSelector patchField = "IngressControllerSelector"
 var IngressControllerCertificate patchField = "IngressControllerCertificate"
 var IngressControllerNodePlacement patchField = "IngressControllerNodePlacement"
+var IngressControllerAnnotation patchField = "ingress.operator.openshift.io/auto-delete-load-balancer"
```
Comment (fahlmant) [inline pkg/controller/publishingstrategy/publishingstrategy_controller.go:47]:
```suggestion
var IngressControllerDeleteLBAnnotation patchField = "ingress.operator.openshift.io/auto-delete-load-balancer"
```

File: pkg/controller/publishingstrategy/publishingstrategy_controller.go [hunk 2]
```diff
@@ -9,7 +9,6 @@ import (
 	"github.com/go-logr/logr"
 	operatorv1 "github.com/openshift/api/operator/v1"
 	"github.com/openshift/cloud-ingress-operator/pkg/apis/cloudingress/v1alpha1"
-	cloudingressv1alpha1 "github.com/openshift/cloud-ingress-operator/pkg/apis/cloudingress/v1alpha1"
```
Comment (boranx) [inline pkg/controller/publishingstrategy/publishingstrategy_controller.go:12]: because it's already defined `v1alpha1` in the previous line.

**Review by fahlmant — COMMENTED** [review]
I believe we should apply this annotation to all IngressControllers that the Publishing Strategy interacts with, which should all have the `owner` annotation present. This should just be default and apps2

**Comment by boranx** [comment]
@fahlmant I've added your suggestion, thanks! Now trying to add unit tests, Btw I saw some race conditions while updating the annotation between _publishing strategy_ and _router_service_

```
{"level":"info","ts":1647519636.549309,"logger":"controller_publishingstrategy","msg":"IngressController's CR of apps2 is being patched for the missing annotation: ingress.operator.openshift.io/auto-delete-load-balancer","Request.Namespace":"openshift-cloud-ingress-operator","Request.Name":"publishingstrategy"}
{"level":"info","ts":1647519636.9850624,"logger":"controller_router_service","msg":"Updating annotation for router-apps2","Request.Namespace":"openshift-ingress","Request.Name":"router-apps2"}
{"level":"error","ts":1647519637.1038272,"logger":"controller_router_service","msg":"Error updating service annotation","Request.Namespace":"openshift-ingress","Request.Name":"router-apps2","error":"Operation cannot be fulfilled on services \"router-apps2\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\tpkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\ngithub.com/openshift/cloud-ingress-operator/pkg/controller/routerservice.(*ReconcileRouterService).Reconcile\n\tsrc/github.com/boranx/cloud-ingress-operator/pkg/controller/routerservice/router_service_controller.go:110\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:298\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\tpkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\tpkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\tpkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\tpkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\tpkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\tpkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:99"}
```

**Comment by ravitri** [comment]
/unhold
Based on discussion with Taylor and James, the intent for this change is understood and the impact is known and expected. Thus, unholding this PR. Thanks James and Taylor

**Comment by boranx** [comment]
now squashed into 1 commit.

**Review by fahlmant — APPROVED** [review]
/lgtm
/approve

**Pull request merged**